File: packages/wasmvm/wasmclient/go/wasmclient/wasmclientevents.go
package wasmclient

import (
	"context"
	"fmt"

	"nhooyr.io/websocket"
	"nhooyr.io/websocket/wsjson"

	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type SubscriptionCommand struct {
	Command string `json:"command"`
	Topic   string `json:"topic"`
}

type Event struct {
	ChainID    wasmtypes.ScChainID
	ContractID wasmtypes.ScHname `json:"contractID"`
	Payload    []byte            `json:"payload"`
	Timestamp  uint64            `json:"timestamp"`
	Topic      string            `json:"topic"`
}

func (e *Event) Bytes() []byte {
	enc := wasmtypes.NewWasmEncoder()
	wasmtypes.HnameEncode(enc, e.ContractID)
	wasmtypes.StringEncode(enc, e.Topic)
	wasmtypes.Uint64Encode(enc, e.Timestamp)
	enc.FixedBytes(e.Payload, uint32(len(e.Payload)))
	return enc.Buf()
}

type ISCEvent struct {
	Kind      string   `json:"kind"`
	Issuer    string   `json:"issuer"`    // (isc.AgentID) nil means issued by the VM
	RequestID string   `json:"requestID"` // (isc.RequestID)
	ChainID   string   `json:"chainID"`   // (isc.ChainID)
	Payload   []*Event `json:"payload"`
}

type WasmClientEvents struct {
	chainID    wasmtypes.ScChainID
	contractID wasmtypes.ScHname
	handler    wasmlib.IEventHandlers
}

func startEventLoop(url string, eventDone chan bool, eventHandlers *[]*WasmClientEvents) error {
	ctx := context.Background()
	ws, _, err := websocket.Dial(ctx, url, nil)
	if err != nil {
		return err
	}
	err = subscribe(ctx, ws, "chains")
	if err != nil {
		return err
	}
	err = subscribe(ctx, ws, "block_events")
	if err != nil {
		return err
	}

	go func() {
		<-eventDone
		_ = ws.Close(websocket.StatusNormalClosure, "intentional close")
	}()

	go eventLoop(ctx, ws, eventHandlers)

	return nil
}

func eventLoop(ctx context.Context, ws *websocket.Conn, eventHandlers *[]*WasmClientEvents) {
	for {
		evt := ISCEvent{}
		err := wsjson.Read(ctx, ws, &evt)
		if err != nil {
			return
		}
		items := evt.Payload
		for _, item := range items {
			event := NewContractEvent(evt.ChainID, item.Bytes())
			for _, h := range *eventHandlers {
				h.ProcessEvent(&event)
			}
		}
	}
}

func NewContractEvent(chainID string, eventData []byte) Event {
	dec := wasmtypes.NewWasmDecoder(eventData)
	hContract := wasmtypes.HnameDecode(dec)
	topic := wasmtypes.StringDecode(dec)
	payload := dec.FixedBytes(dec.Length())
	event := Event{
		ChainID:    wasmtypes.ChainIDFromString(chainID),
		ContractID: hContract,
		Payload:    payload,
		Timestamp:  wasmtypes.Uint64FromBytes(payload[:wasmtypes.ScUint64Length]),
		Topic:      topic,
	}
	return event
}

func (h WasmClientEvents) ProcessEvent(event *Event) {
	if event.ContractID != h.contractID || event.ChainID != h.chainID {
		return
	}
	fmt.Printf("%s %s %s\n", event.ChainID.String(), event.ContractID.String(), event.Topic)
	dec := wasmtypes.NewWasmDecoder(event.Payload)
	h.handler.CallHandler(event.Topic, dec)
}

func RemoveHandler(eventHandlers []*WasmClientEvents, eventsID uint32) []*WasmClientEvents {
	eh := eventHandlers[:0]
	for _, h := range eventHandlers {
		if h.handler.ID() != eventsID {
			eh = append(eh, h)
		}
	}
	return eh
}

func subscribe(ctx context.Context, ws *websocket.Conn, topic string) error {
	msg := SubscriptionCommand{
		Command: "subscribe",
		Topic:   topic,
	}
	err := wsjson.Write(ctx, ws, msg)
	if err != nil {
		return err
	}
	return wsjson.Read(ctx, ws, &msg)
}

File: packages/wasmvm/wasmclient/ts/wasmclient/lib/wasmclientevents.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {Converter} from '@iota/util.js';
import * as iscclient from './iscclient';
import * as wasmlib from 'wasmlib';
import {RawData, WebSocket} from 'ws';

export class Event {
    chainID: wasmlib.ScChainID;
    contractID: wasmlib.ScHname;
    payload: Uint8Array;
    timestamp: string;
    topic: string;

    public constructor(chainID: string, event: any) {
        this.chainID = wasmlib.chainIDFromString(chainID);
        this.contractID = new wasmlib.ScHname(event.contractID);
        this.topic = event.topic;
        this.timestamp = event.timestamp.toString();
        const timestamp = wasmlib.uint64FromString(this.timestamp);
        this.payload = wasmlib.concat(wasmlib.uint64ToBytes(timestamp), Converter.base64ToBytes(event.payload));
    }
}

export class WasmClientEvents {
    chainID: wasmlib.ScChainID;
    contractID: wasmlib.ScHname;
    handler: wasmlib.IEventHandlers;

    constructor(chainID: wasmlib.ScChainID, contractID: wasmlib.ScHname, handler: wasmlib.IEventHandlers) {
        this.chainID = chainID;
        this.contractID = contractID;
        this.handler = handler;
    }

    public static startEventLoop(ws: WebSocket, eventHandlers: WasmClientEvents[]): iscclient.Error {
        ws.on('open', () => {
            this.subscribe(ws, 'chains');
            this.subscribe(ws, 'block_events');
        });
        ws.on('error', (err) => {
            // callback(['error', err.toString()]);
        });
        ws.on('message', (data) => this.eventLoop(data, eventHandlers));
        return null;
    }

    private static eventLoop(data: RawData, eventHandlers: WasmClientEvents[]) {
        let msg: any;
        try {
            const json = data.toString();
            console.log(json);
            msg = JSON.parse(json);
            if (!msg.kind) {
                // filter out subscribe responses
                return;
            }
            console.log(msg);
        } catch (ex) {
            console.log(`Failed to parse expected JSON message: ${data} ${ex}`);
            return;
        }

        for (const item of msg.payload) {
            const event = new Event(msg.chainID, item);
            for (const h of eventHandlers) {
                h.processEvent(event);
            }
        }
    }

    private static subscribe(ws: WebSocket, topic: string) {
        const msg = {
            command: 'subscribe',
            topic: topic,
        };
        const rawMsg = JSON.stringify(msg);
        ws.send(rawMsg);
    }

    private processEvent(event: Event) {
        if (!event.contractID.equals(this.contractID) || !event.chainID.equals(this.chainID)) {
            return;
        }
        console.log(event.chainID.toString() + ' ' + event.contractID.toString() + ' ' + event.topic);
        const dec = new wasmlib.WasmDecoder(event.payload);
        this.handler.callHandler(event.topic, dec);
    }
}
